Numpy 2: OverflowError with int96

**Describe the issue**:

Upgrading from numpy  1 to 2.0.0rc1, the fastparquet test suite starts to fail during the unit tests for the openSUSE rpm packaging builds.

```python
[   76s] _______________________________ test_int96_stats _______________________________
[   76s] [gw2] linux -- Python 3.10.14 /usr/bin/python3.10
[   76s] 
[   76s] tempdir = '/tmp/tmpmti7pznk'
[   76s] 
[   76s]     def test_int96_stats(tempdir):
[   76s]         df = makeMixedDataFrame()
[   76s]     
[   76s]         fn = os.path.join(tempdir, 'foo.parquet')
[   76s]         write(fn, df, row_group_offsets=[0, 2], times='int96')
[   76s]     
[   76s]         p = ParquetFile(fn)
[   76s]     
[   76s] >       s = statistics(p)
[   76s] 
[   76s] ../../BUILDROOT/python-fastparquet-2024.2.0-63.1.x86_64/usr/lib64/python3.10/site-packages/fastparquet/test/test_api.py:892: 
[   76s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[   76s] ../../BUILDROOT/python-fastparquet-2024.2.0-63.1.x86_64/usr/lib64/python3.10/site-packages/fastparquet/api.py:1278: in statistics
[   76s]     else list(converted_types.convert(
[   76s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[   76s] 
[   76s] data = array([b'\x00\x00\x00\x00\x00\x00\x00\x001u%',
[   76s]        b'\x00\x00\x00\x00\x00\x00\x00\x005u%'], dtype='|S12')
[   76s] se = {'type': 3, 'type_length': None, 'repetition_type': 1, 'name': 'D', 'num_children': None, 'converted_type': None, 'scale': None, 'precision': None, 'field_id': None, 'logicalType': None}
[   76s] timestamp96 = True, dtype = None
[   76s] 
[   76s]     def convert(data, se, timestamp96=True, dtype=None):
[   76s]         """Convert known types from primitive to rich.
[   76s]     
[   76s]         Parameters
[   76s]         ----------
[   76s]         data: pandas series of primitive type
[   76s]         se: a schema element.
[   76s]         timestamp96: convert int96 as if it were written by mr-parquet
[   76s]         """
[   76s]         ctype = se.converted_type
[   76s]         if se.type == parquet_thrift.Type.INT96 and timestamp96:
[   76s]             data2 = data.view([('ns', 'i8'), ('day', 'i4')])
[   76s]             # TODO: this should be ms unit, now that we can?
[   76s] >           return ((data2['day'] - 2440588) * 86400000000000 +
[   76s]                     data2['ns']).view('M8[ns]')
[   76s] E           OverflowError: Python integer 86400000000000 out of bounds for int32
[   76s] 
[   76s] ../../BUILDROOT/python-fastparquet-2024.2.0-63.1.x86_64/usr/lib64/python3.10/site-packages/fastparquet/converted_types.py:161: OverflowError
...
[   76s] =========================== short test summary info ============================
[   76s] FAILED test/test_api.py::test_int96_stats - OverflowError: Python integer 864...
[   76s] FAILED test/test_api.py::test_spark_date_empty_rg - OverflowError: Python int...
[   76s] FAILED test/test_read.py::test_timestamp96 - OverflowError: Python integer 86...
[   76s] FAILED test/test_read.py::test_truncated_decimal - OverflowError: Python inte...
[   76s] ===== 4 failed, 380 passed, 123 skipped, 1 xfailed, 26 warnings in 20.34s ======
[   76s] error: Bad exit status from /var/tmp/rpm-tmp.QIGekb (%check)
```

Full build log: 

[fastparquet-numpy2.txt](https://github.com/dask/fastparquet/files/15174794/fastparquet-numpy2.txt)

**Minimal Complete Verifiable Example**:

- get all dependencies compiled with numpy 2
- run test suite

**Environment**:

- Dask version: 2024.4.2
- Python version: All of them
- Operating System: openSUSE Tumbleweed
- Install method (conda, pip, source): source / rpm packages from source

--- a/.github/workflows/main.yaml
+++ b/.github/workflows/main.yaml
@@ -70,36 +70,6 @@ jobs:
         run: |
           echo "FASTPARQUET_DATAPAGE_V2=$FASTPARQUET_DATAPAGE_V2"
           pytest --verbose --cov=fastparquet
-  dask:
-    name: dask
-    runs-on: ubuntu-latest
-    steps:
-      - name: APT
-        run: sudo apt-get install liblzo2-dev
-
-      - name: Checkout
-        uses: actions/checkout@v4
-        with:
-          fetch-depth: 0
-
-      - name: Setup conda
-        uses: mamba-org/provision-with-micromamba@main
-        with:
-          environment-file: ci/environment-py39.yml
-
-      - name: pip-install
-        shell: bash -l {0}
-        run: |
-          git clone https://github.com/dask/dask
-          pip install pyarrow
-          pip install -e dask/
-          pip install -e . --no-deps
-
-      - name: Run Tests
-        shell: bash -l {0}
-        run: |
-          pytest --verbose dask/dask/dataframe/io/tests/test_parquet.py
-  
   pandas:
     name: pandas
     runs-on: ubuntu-latest

--- a/fastparquet/converted_types.py
+++ b/fastparquet/converted_types.py
@@ -31,8 +31,9 @@ except ImportError:  # pragma: no cover
         def tobson(x):
             raise ImportError("BSON not found")
 
-DAYS_TO_MILLIS = 86400000000000
-"""Number of millis in a day. Used to convert a Date to a date"""
+# Explicitly use numpy type in order to avoid promotion errors due to NEP 50 in numpy >= 2
+DAYS_TO_NANOS = np.int64(86400000000000)
+"""Number of nanoseconds in a day. Used to convert a Date to a date"""
 nat = np.datetime64('NaT').view('int64')
 
 simple = {
@@ -158,7 +159,7 @@ def convert(data, se, timestamp96=True, dtype=None):
     if se.type == parquet_thrift.Type.INT96 and timestamp96:
         data2 = data.view([('ns', 'i8'), ('day', 'i4')])
         # TODO: this should be ms unit, now that we can?
-        return ((data2['day'] - 2440588) * 86400000000000 +
+        return ((data2['day'] - np.int64(2440588)) * DAYS_TO_NANOS +
                 data2['ns']).view('M8[ns]')
     if se.logicalType is not None and se.logicalType.TIMESTAMP is not None:
         dt = _logical_to_time_dtype(se.logicalType.TIMESTAMP)
@@ -188,7 +189,7 @@ def convert(data, se, timestamp96=True, dtype=None):
                 for i in range(len(data))
             ])
     elif ctype == parquet_thrift.ConvertedType.DATE:
-        data = data * DAYS_TO_MILLIS
+        data = data * DAYS_TO_NANOS
         return data.view('datetime64[ns]')
     elif ctype == parquet_thrift.ConvertedType.TIME_MILLIS:
         # this was not covered by new pandas time units

--- a/fastparquet/dataframe.py
+++ b/fastparquet/dataframe.py
@@ -153,7 +153,7 @@ def empty(types, size, cats=None, cols=None, index_types=None, index_names=None,
                     tz_to_dt_tz(timezones[str(col)]))
             else:
                 index = Index(d)
-            views[col] = index.values
+            views[col] = d
     else:
         index = MultiIndex([[]], [[]])
         # index = MultiIndex.from_arrays(indexes)

--- a/fastparquet/test/test_output.py
+++ b/fastparquet/test/test_output.py
@@ -1022,9 +1022,7 @@ def test_no_string(tmpdir):
     df["A"] = df["A"].astype(pd.StringDtype())
 
     # set *all* values to NA
-    df["A"].iloc[0] = pd.NA
-    df["A"].iloc[1] = pd.NA
-    df["A"].iloc[2] = pd.NA
+    df.loc[:, "A"] = pd.NA
     df.to_parquet(fn, engine="fastparquet")
     df2 = pd.read_parquet(fn)
     assert pd.isna(df2.A).all()

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,2 +1,2 @@
 [build-system]
-requires = ["setuptools", "wheel", "Cython >= 0.29.23", "oldest-supported-numpy", "pytest-runner"]
+requires = ["setuptools", "setuptools_scm", "Cython >= 0.29.23", "numpy>=2.0.0rc1"]

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,5 +1,5 @@
 pandas>=1.5.0
-numpy>=1.20.3
+numpy
 cramjam>=2.3
 fsspec
 packaging

--- a/setup.py
+++ b/setup.py
@@ -53,13 +53,6 @@ setup(
         'local_scheme': 'no-local-version',
         'write_to': 'fastparquet/_version.py'
     },
-    setup_requires=[
-        'setuptools>18.0',
-        'setuptools-scm>1.5.4',
-        'Cython',
-        'pytest-runner',
-        'oldest-supported-numpy'
-    ],
     description='Python support for Parquet file format',
     author='Martin Durant',
     author_email='mdurant@anaconda.com',

